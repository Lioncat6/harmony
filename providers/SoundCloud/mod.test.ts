import { describeProvider, makeProviderOptions } from '@/providers/test_spec.ts';
import { stubProviderLookups } from '@/providers/test_stubs.ts';
import { assert } from 'std/assert/assert.ts';
import { afterAll, describe } from '@std/testing/bdd';
import { assertSnapshot } from '@std/testing/snapshot';

import SoundCloudProvider from './mod.ts';

describe('SoundCloud provider', () => {
	const sc = new SoundCloudProvider(makeProviderOptions());
	const lookupStub = stubProviderLookups(sc);

	describeProvider(sc, {
		urls: [{
			description: 'album page',
			url: new URL('https://soundcloud.com/ivycomb/sets/crimsongalaxies'),
			id: { type: 'playlist', id: 'ivycomb/crimsongalaxies' },
			isCanonical: true,
		}, {
			description: 'standalone track page',
			url: new URL('https://soundcloud.com/lonealphamusic/magazines'),
			id: { type: 'track', id: 'lonealphamusic/magazines' },
			serializedId: 'lonealphamusic/track/magazines',
			isCanonical: true,
		}, {
			description: 'artist page',
			url: new URL('https://soundcloud.com/vocalokat'),
			id: { type: 'artist', id: 'vocalokat' },
			isCanonical: true,
		}],
		releaseLookup: [{
			description: 'track release with downloads enabled',
			release: 'leagueoflegends/track/piercing-light-mako-remix',
			assert: async (release, ctx) => {
				await assertSnapshot(ctx, release);
				const isFree = release.externalLinks.some((link) => link.types?.includes('free download'));
				assert(isFree, 'Release should be downloadable for free');
			},
		}],
	});

	afterAll(() => {
		lookupStub.restore();
	});
});
