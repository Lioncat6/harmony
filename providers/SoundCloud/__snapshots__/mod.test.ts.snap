export const snapshot = {};

snapshot[`SoundCloud provider > release lookup > track release with downloads enabled 1`] = `
{
  artists: [
    {
      creditedName: "League of Legends",
      externalIds: [
        {
          id: "leagueoflegends",
          provider: "soundcloud",
          type: "artist",
        },
      ],
      name: "League of Legends",
    },
  ],
  availableIn: undefined,
  externalLinks: [
    {
      types: [
        "free streaming",
        "free download",
      ],
      url: "https://soundcloud.com/leagueoflegends/piercing-light-mako-remix",
    },
  ],
  images: [
    {
      provider: "SoundCloud",
      thumbUrl: "https://i1.sndcdn.com/artworks-000142912000-f05col-t300x300.jpg",
      types: [
        "front",
      ],
      url: "https://i1.sndcdn.com/artworks-000142912000-f05col-original.jpg",
    },
  ],
  info: {
    messages: [],
    providers: [
      {
        apiUrl: undefined,
        id: "leagueoflegends/track/piercing-light-mako-remix",
        internalName: "soundcloud",
        lookup: {
          method: "id",
          value: "leagueoflegends/track/piercing-light-mako-remix",
        },
        name: "SoundCloud",
        url: "https://soundcloud.com/leagueoflegends/piercing-light-mako-remix",
      },
    ],
  },
  labels: undefined,
  media: [
    {
      format: "Digital Media",
      tracklist: [
        {
          artists: [
            {
              creditedName: "League of Legends",
              externalIds: [
                {
                  id: "leagueoflegends",
                  provider: "soundcloud",
                  type: "artist",
                },
              ],
              name: "League of Legends",
            },
          ],
          availableIn: undefined,
          isrc: undefined,
          length: 291422,
          number: 1,
          recording: {
            externalIds: [
              {
                id: "leagueoflegends/piercing-light-mako-remix",
                provider: "soundcloud",
                type: "track",
              },
            ],
          },
          title: "Piercing Light (Mako Remix)",
        },
      ],
    },
  ],
  packaging: "None",
  releaseDate: {
    day: 12,
    month: 1,
    year: 2016,
  },
  title: "Piercing Light (Mako Remix)",
  types: [
    "Single",
  ],
}
`;
